pkg/push/api/api.go

```diff
@@ -631,13 +621,79 @@ func (s *PushService) containsTags(
 	return nil
 }
 
-func (s *PushService) containsFCMKey(ctx context.Context, pushes []*pushproto.Push, fcmAPIKey string) bool {
+func (s *PushService) checkFCMServiceAccount(
+	ctx context.Context,
+	pushes []*pushproto.Push,
+	fcmServiceAccount []byte,
+	localizer locale.Localizer,
+) error {
+	// Check if the JSON is a service account file
+	_, err := google.CredentialsFromJSON(
+		ctx,
+		[]byte(fcmServiceAccount),
+		"https://www.googleapis.com/auth/firebase.messaging",
+	)
+	if err != nil {
+		s.logger.Error("failed to get credentials from JSON", zap.Error(err))
+		dt, err := statusFCMServiceAccountInvalid.WithDetails(&errdetails.LocalizedMessage{
+			Locale:  localizer.GetLocale(),
+			Message: localizer.MustLocalizeWithTemplate(locale.InvalidArgumentError, "FCM service account"),
+		})
+		if err != nil {
+			return statusInternal.Err()
+		}
+		return dt.Err()
+	}
+	// Check if the service account already exists in the database
 	for _, push := range pushes {
-		if push.FcmApiKey == fcmAPIKey {
-			return true
+		equal, err := s.compareJSON(push.FcmServiceAccount, string(fcmServiceAccount))
+		if err != nil {
+			s.logger.Error("failed to compare the JSON", zap.Error(err))
+			dt, err := statusInternal.WithDetails(&errdetails.LocalizedMessage{
+				Locale:  localizer.GetLocale(),
+				Message: localizer.MustLocalize(locale.InternalServerError),
+			})
+			if err != nil {
+				return statusInternal.Err()
+			}
+			return dt.Err()
+		}
+		if equal {
+			s.logger.Error("fcm service account already exists in the database")
+			dt, err := statusFCMServiceAccountAlreadyExists.WithDetails(&errdetails.LocalizedMessage{
+				Locale:  localizer.GetLocale(),
+				Message: localizer.MustLocalizeWithTemplate(locale.AlreadyExistsError, "FCM service account"),
+			})
+			if err != nil {
+				return statusInternal.Err()
+			}
+			return dt.Err()
 		}
 	}
-	return false
+	return nil
+}
+
+// compareJSON compares two JSON strings and returns true if they are equivalent
+func (s *PushService) compareJSON(jsonStr1, jsonStr2 string) (bool, error) {
+	var obj1, obj2 interface{}
+	// Unmarshal the JSON strings into Go data structures
+	if err := json.Unmarshal([]byte(jsonStr1), &obj1); err != nil {
+		return false, err
+	}
+	if err := json.Unmarshal([]byte(jsonStr2), &obj2); err != nil {
+		return false, err
+	}
+	// Marshal the Go data structures into canonical JSON format
+	json1, err := json.Marshal(obj1)
+	if err != nil {
+		return false, err
+	}
+	json2, err := json.Marshal(obj2)
+	if err != nil {
+		return false, err
+	}
+	// Compare the canonical JSON representations
+	return bytes.Equal(json1, json2), nil
 }
```

Comment:
For comparison purpose, we can consider use `json.RawMessage` instead of `interface{}`.

Advantages:
- This approach is more efficient as it avoids fully parsing the JSON into Go data structures.
- It reduces the number of allocations and copying operations.

```go
// compareJSON compares two JSON strings and returns true if they are equivalent
func (s *PushService) compareJSON(jsonStr1, jsonStr2 string) (bool, error) {
    var raw1, raw2 json.RawMessage

    // Unmarshal the JSON strings into json.RawMessage
    if err := json.Unmarshal([]byte(jsonStr1), &raw1); err != nil {
        return false, err
    }
    if err := json.Unmarshal([]byte(jsonStr2), &raw2); err != nil {
        return false, err
    }

    // Marshal the json.RawMessage back into JSON for normalization
    normalized1, err := json.Marshal(raw1)
    if err != nil {
        return false, err
    }
    normalized2, err := json.Marshal(raw2)
    if err != nil {
        return false, err
    }

    // Compare the normalized JSON representations
    return bytes.Equal(normalized1, normalized2), nil
}
```